reactor-core/src/main/java/reactor/core/scheduler/BoundedElasticScheduler.java

```diff
@@ -402,30 +472,45 @@ else if (a < parent.maxThreads) {
 						ScheduledExecutorService s = Schedulers.decorateExecutorService(parent, parent.createBoundedExecutorService());
 						BoundedState newState = new BoundedState(this, s);
 						if (newState.markPicked()) {
-							busyQueue.add(newState);
+							setBusy(newState);
 							return newState;
 						}
 					}
 					//else optimistically retry (implicit continue here)
 				}
 				else {
-					//pick the least busy one
-					BoundedState s = busyQueue.poll();
+					BoundedState s = choseOneBusy();
 					if (s != null && s.markPicked()) {
-						busyQueue.add(s); //put it back in the queue with updated priority
 						return s;
 					}
 					//else optimistically retry (implicit continue here)
 				}
 			}
 		}
 
-		void setIdle(BoundedState boundedState) {
-			//impl. note: reversed order could lead to a race condition where state is added to idleQueue
-			//then concurrently pick()ed into busyQueue then removed from same busyQueue.
-			if (this.busyQueue.remove(boundedState)) {
-				this.idleQueue.add(boundedState);
+		@Nullable
+		private BoundedState choseOneBusy() {
+			BoundedState[] arr = busyArray;
+			int len = arr.length;
+			if (len == 0) {
+				return null; //implicit retry in the pick() loop
 			}
+			if (len == 1) {
+				return arr[0];
+			}
+
+			BoundedState choice = arr[0];
+			int leastBusy = Integer.MAX_VALUE;
+
+			for (int i = 0; i < arr.length; i++) {
```

Comment:
can we do random choices here instead of iterating?

Reply:
that is what we do in rsocket weighted loadbalancer -> https://github.com/rsocket/rsocket-java/blob/master/rsocket-core/src/main/java/io/rsocket/loadbalance/WeightedLoadbalanceStrategy.java#L91

Reply:
let me capture that as a follow-up issue and merge this one for now 👍

Reply:
see #2927